File: Muselly.Web/wwwroot/interop.js
// Browser interop for the Muselly web client: the page origin (for the API base URL) and a single shared
// HTMLAudioElement that streams Opus from the host's /api/stream endpoint. The managed BrowserAudio service
// polls position/duration/ended through these functions.

let audio = null;
let ended = false;

export function getOrigin() {
    return globalThis.location.origin;
}

export function getPath() {
    return globalThis.location.pathname + globalThis.location.search;
}

export function setPath(path) {
    try { globalThis.history.replaceState(null, "", path); } catch (e) { /* ignore */ }
}

export function audioInit() {
    if (audio) return;
    audio = new Audio();
    audio.preload = "auto";
    audio.addEventListener("ended", () => { ended = true; });
    audio.addEventListener("playing", () => { ended = false; });
}

export function audioPlay(url) {
    audioInit();
    ended = false;
    audio.src = url;
    const p = audio.play();
    if (p && p.catch) p.catch(() => {});
}

export function audioPause() {
    if (audio) audio.pause();
}

export function audioResume() {
    if (audio) {
        const p = audio.play();
        if (p && p.catch) p.catch(() => {});
    }
}

export function audioStop() {
    if (audio) {
        audio.pause();
        audio.removeAttribute("src");
        audio.load();
    }
    ended = false;
}

export function audioSeek(seconds) {
    if (audio) {
        try { audio.currentTime = seconds; } catch (e) { /* not seekable yet */ }
    }
}

export function audioSetVolume(v) {
    if (audio) audio.volume = v;
}

export function audioGetTime() {
    return audio ? (audio.currentTime || 0) : 0;
}

export function audioGetDuration() {
    return (audio && isFinite(audio.duration)) ? audio.duration : 0;
}

export function audioEnded() {
    return ended;
}

File: Muselly.Web/Audio/Interop.cs
using System.Runtime.InteropServices.JavaScript;
using System.Runtime.Versioning;
using System.Threading.Tasks;

namespace Muselly.Web.Audio;

/// <summary>
/// Managed bindings to <c>wwwroot/interop.js</c>: the page origin and the shared HTMLAudioElement used to
/// stream Opus from the host. <see cref="InitAsync"/> must complete before any other member is called.
/// </summary>
[SupportedOSPlatform("browser")]
internal static partial class Interop
{
    // The runtime resolves module URLs relative to the _framework/ directory, but interop.js is deployed at
    // the app root (next to index.html), so step up one level.
    public static Task InitAsync() => JSHost.ImportAsync("interop", "../interop.js");

    [JSImport("getOrigin", "interop")] public static partial string GetOrigin();
    [JSImport("getPath", "interop")] public static partial string GetPath();
    [JSImport("setPath", "interop")] public static partial void SetPath(string path);

    [JSImport("audioInit", "interop")] public static partial void AudioInit();
    [JSImport("audioPlay", "interop")] public static partial void AudioPlay(string url);
    [JSImport("audioPause", "interop")] public static partial void AudioPause();
    [JSImport("audioResume", "interop")] public static partial void AudioResume();
    [JSImport("audioStop", "interop")] public static partial void AudioStop();
    [JSImport("audioSeek", "interop")] public static partial void AudioSeek(double seconds);
    [JSImport("audioSetVolume", "interop")] public static partial void AudioSetVolume(double volume);
    [JSImport("audioGetTime", "interop")] public static partial double AudioGetTime();
    [JSImport("audioGetDuration", "interop")] public static partial double AudioGetDuration();
    [JSImport("audioEnded", "interop")] public static partial bool AudioEnded();
}
